apps/system/js/sync_state_machine.js

```diff
@@ -0,0 +1,147 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/*
+
+       +---------------Success ---------------+
+       |                                      |
+   +---+----+                                 |
+   |Enabling+---Error---+                     |
+   +---+----+           |                     |
+       ^              +-v---+                 |
+       |   +-Disable--+Error+-------------+   |
+       |   |          +--+--+             |   |
+     Enable|             ^             Enable |
+       |   |             |                |   |
+       |   |             |                |   |
+  +----+---v-+           |            +---v---v+
+  | Disabled |         Error +------->+Enabled +---+
+  +----^-----+           |   |        +---+----+   |
+       |                 |   |            |        |
+       |                 | Success       Sync      |
+       |                 |   |            |        |
+       |             +---+---+            |        |
+       |             |Syncing<------------+        |
+       |             +-------+                     |
+       |                                           |
+       +--------------Disable  --------------------+
+
+*/
+
+'use strict';
+
+(function(exports) {
+
+  var SyncStateMachine = {
+    current: 'disabled'
+  };
+
+  /**
+   * Object representing the possible events and the associated states.
+   * {
+   *   eventName: {
+   *     from: to,
+   *     from: to
+   *     ...
+   *   },
+   *   ...
+   * }
+   */
+  const events = {
+    enable: {
+      'disabled': 'enabling',
+      'error': 'enabled'
```

Comment:
The `enabling` handler will be the one handling the FxA login and all the other things (if any) we need to do to actually enable sync. If we come from a sync error, that means that we were already enabled, we don't need to do anything with FxA again or any other "enabling" task, so we can go to `enable` directly.